From 0f46d701bb4d6d81106e9a5a0a13e7237d311efa Mon Sep 17 00:00:00 2001
From: Tony Wickham <twickham@google.com>
Date: Fri, 22 Aug 2025 12:10:30 -0700
Subject: [PATCH] Add onKeyEvent, to support KEYCODE_HOME

- Also added displayId to onBackEvent

Bug: 381339822
Flag: com.android.launcher3.home_button_uses_keycode_home
Test: Manual, OverviewCommandHelperTest
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:622a882e1f17d108a336ecfd3e4eb7aa9707f1e6)
Merged-In: I55e2ee7adfa40c9d8f289a2b4c76aa12f80e78c3
Change-Id: I55e2ee7adfa40c9d8f289a2b4c76aa12f80e78c3
---
 .../shared/recents/ISystemUiProxy.aidl        | 12 ++++++++---
 .../recents/LauncherProxyService.java         | 21 ++++++++++++-------
 2 files changed, 23 insertions(+), 10 deletions(-)

diff --git a/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl b/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl
index 1f6bea18d53a..bf01826dc0c5 100644
--- a/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl
+++ b/packages/SystemUI/shared/src/com/android/systemui/shared/recents/ISystemUiProxy.aidl
@@ -103,9 +103,10 @@ interface ISystemUiProxy {
     oneway void expandNotificationPanel() = 29;
 
     /**
-     * Notifies SystemUI of a back KeyEvent.
+     * Notifies SystemUI of KEYCODE_BACK. If the passed KeyEvent is not null, SystemUI may use it to
+     * show a predictive back animation, otherwise it will send a ACTION_DOWN followed by ACTION_UP.
      */
-    oneway void onBackEvent(in KeyEvent keyEvent) = 44;
+    oneway void onBackEvent(in KeyEvent keyEvent, int displayId) = 44;
 
     /** Sets home rotation enabled. */
     oneway void setHomeRotationEnabled(boolean enabled) = 45;
@@ -178,5 +179,10 @@ interface ISystemUiProxy {
      */
     oneway void updateContextualEduStats(boolean isTrackpadGesture, String gestureType) = 58;
 
-    // Next id = 59
+    /**
+     * Notifies SystemUI of a KeyEvent of the specified type (e.g. KEYCODE_BACK, KEYCODE_HOME).
+     */
+    oneway void onKeyEvent(int keycode, int displayId) = 60;
+
+    // Next id = 61
 }
diff --git a/packages/SystemUI/src/com/android/systemui/recents/LauncherProxyService.java b/packages/SystemUI/src/com/android/systemui/recents/LauncherProxyService.java
index 5b9717535ab5..3d0127c934e7 100644
--- a/packages/SystemUI/src/com/android/systemui/recents/LauncherProxyService.java
+++ b/packages/SystemUI/src/com/android/systemui/recents/LauncherProxyService.java
@@ -302,20 +302,27 @@ public class LauncherProxyService implements CallbackController<LauncherProxyLis
         }
 
         @Override
-        public void onBackEvent(@Nullable KeyEvent keyEvent) throws RemoteException {
+        public void onBackEvent(@Nullable KeyEvent keyEvent, int displayId) throws RemoteException {
             if (predictiveBackThreeButtonNav() && predictiveBackSwipeEdgeNoneApi()
                     && mBackAnimation != null && keyEvent != null) {
                 mBackAnimation.setTriggerBack(!keyEvent.isCanceled());
                 mBackAnimation.onBackMotion(/* touchX */ 0, /* touchY */ 0, keyEvent.getAction(),
                         EDGE_NONE);
             } else {
-                verifyCallerAndClearCallingIdentityPostMain("onBackPressed", () -> {
-                    sendEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_BACK);
-                    sendEvent(KeyEvent.ACTION_UP, KeyEvent.KEYCODE_BACK);
-                });
+                onKeyEvent(KeyEvent.KEYCODE_BACK, displayId);
             }
         }
 
+        @Override
+        public void onKeyEvent(int keycode, int displayId) {
+            verifyCallerAndClearCallingIdentityPostMain(
+                    "onKeyEvent " + KeyEvent.keyCodeToString(keycode) + " displayId=" + displayId,
+                    () -> {
+                        sendEvent(KeyEvent.ACTION_DOWN, keycode, displayId);
+                        sendEvent(KeyEvent.ACTION_UP, keycode, displayId);
+                    });
+        }
+
         @Override
         public void onImeSwitcherPressed() {
             // TODO(b/204901476) We're intentionally using the default display for now since
@@ -370,14 +377,14 @@ public class LauncherProxyService implements CallbackController<LauncherProxyLis
                     onTaskbarAutohideSuspend(suspend));
         }
 
-        private boolean sendEvent(int action, int code) {
+        private boolean sendEvent(int action, int code, int displayId) {
             long when = SystemClock.uptimeMillis();
             final KeyEvent ev = new KeyEvent(when, when, action, code, 0 /* repeat */,
                     0 /* metaState */, KeyCharacterMap.VIRTUAL_KEYBOARD, 0 /* scancode */,
                     KeyEvent.FLAG_FROM_SYSTEM | KeyEvent.FLAG_VIRTUAL_HARD_KEY,
                     InputDevice.SOURCE_KEYBOARD);
 
-            ev.setDisplayId(mContext.getDisplay().getDisplayId());
+            ev.setDisplayId(displayId);
             return InputManagerGlobal.getInstance()
                     .injectInputEvent(ev, InputManager.INJECT_INPUT_EVENT_MODE_ASYNC);
         }
-- 
2.52.0.rc1.455.g30608eb744-goog

